File: data_structure_python/heap/min_heap.py
import random


class MinHeap(object):
    def __init__(self):
        self.data = []

    def __len__(self):
        return len(self.data)

    def is_empty(self):
        return len(self.data) == 0

    def _parent(self, index):
        if index == 0:
            raise IndexError("index 0 have no parent")
        return (index - 1) // 2

    def _left_child(self, index):
        return index * 2 + 1

    def _right_child(self, index):
        return index * 2 + 2

    def add(self, val):
        self.data.append(val)
        self._sift_up(len(self.data) - 1)

    def _sift_up(self, k):
        while k > 0 and self.data[self._parent(k)] > self.data[k]:
            self.data[self._parent(k)], self.data[k] = self.data[k], self.data[self._parent(k)]
            k = self._parent(k)

    def find_min(self):
        if len(self.data) == 0:
            raise ValueError("can't find min in empty heap")
        return self.data[0]

    def extract_min(self):
        ret = self.find_min()
        self.data[0], self.data[len(self.data) - 1] = self.data[len(self.data) - 1], self.data[0]
        self.data.pop()
        self._sift_down(0)
        return ret

    def _sift_down(self, k):
        while self._left_child(k) < len(self.data):
            i = self._left_child(k)
            if i + 1 < len(self.data) and self.data[i + 1] < self.data[i]:
                i += 1
                # 此时i 是左右孩子里的那个最小的
            if self.data[k] <= self.data[i]:
                break
            self.data[k], self.data[i] = self.data[i], self.data[k]
            k = i

    # 替换队顶元素为val 并返回堆顶元素
    def replace(self, val):
        ret = self.find_min()
        self.data[0] = val
        self._sift_down(0)
        return ret

    # 从第一个非叶子节点开始，从后向前进行 sift down操作
    def heapify(self, arr):
        self.data = arr.copy()
        if len(arr) != 1:
            # 最后一个非叶子节点 ： 最后一个节点的父亲节点
            i = self._parent(len(self.data) - 1)
            while i >= 0:
                self._sift_down(i)
                i -= 1

    def __str__(self):
        return str(self.data)


if __name__ == '__main__':
    hq = MinHeap()
    arr1 = [4, 6, 2, 3, 9]
    hq.heapify(arr1)
    print(hq)

File: DataStructureJava/array/heap/MinHeap.java
import java.util.Random;

public class MinHeap<E extends Comparable<E>> {
    private Array<E> data;

    public MinHeap(int capacity) {
        data = new Array<>(capacity);
    }

    public MinHeap() {
        data = new Array<>();
    }

    public MinHeap(E[] arr) {
        data = new Array<>(arr);
        if (arr.length != 1) {
            for (int i = parent(arr.length - 1); i >= 0; i--)
                siftDown(i);
        }
    }

    public int getSize() {
        return data.getSize();
    }

    public boolean isEmpty() {
        return data.isEmpty();
    }

    // 返回完全二叉树的数组表示中，一个索引所表示的元素的父亲节点的索引
    private int parent(int index) {
        if (index == 0) {
            throw new IllegalArgumentException("index - 0 doesn't have parent");
        }
        return (index - 1) / 2;
    }

    // 返回完全二叉树的数组表示中，一个索引所表示的元素的左孩子节点的索引
    private int leftChild(int index) {
        return index * 2 + 1;
    }

    // 返回完全二叉树的数组表示中，一个索引所表示的元素的右孩子节点的索引
    private int rightChild(int index) {
        return index * 2 + 2;
    }

    public void add(E e) {
        data.addLast(e);
        siftUp(data.getSize() - 1);
    }

    private void siftUp(int k) {
        while (k > 0 && data.get(parent(k)).compareTo(data.get(k)) > 0) {
            data.swap(k, parent(k));
            k = parent(k);
        }
    }

    public E findMin() {
        if (data.getSize() == 0) {
            throw new IllegalArgumentException("Can't find min when heap is empty");
        }
        return data.get(0);
    }

    // 取出堆中最小元素
    public E extractMin() {
        E ret = findMin();
        data.swap(0, data.getSize() - 1);
        data.removeLast();
        siftDown(0);
        return ret;
    }

    private void siftDown(int k) {
        while (leftChild(k) < data.getSize()) {
            int j = leftChild(k);
            // 如果右孩子的值小于左孩子的值，
            if (j + 1 < data.getSize() && data.get(j + 1).compareTo(data.get(j)) < 0)
                j++; // j = rightChild(k)
            // 此时 j 是左右孩子中的最小值
            if (data.get(k).compareTo(data.get(j)) <= 0)
                break;
            data.swap(k, j);
            k = j;
        }
    }

    //  取出堆中最小的元素，并且替换成元素e
    public E replace(E e) {
        E ret = findMin();
        data.set(0, e);
        siftDown(0);
        return ret;
    }

    public static void main(String[] args) {
        int n = 10000;
        MinHeap<Integer> minHeap = new MinHeap<>();
        Random rnd = new Random();
        for(int i = 0; i < n; i ++)
            minHeap.add(rnd.nextInt(Integer.MAX_VALUE));

        int[] arr = new int[n];
        for (int i = 0; i < n; i++)
            arr[i] = minHeap.extractMin();
        for (int i = 1; i < n; i++)
            if (arr[i - 1] > arr[i])
                throw new IllegalArgumentException("Error");

        System.out.println("finish");
    }
}
